### PR TITLE
SCC-3776 - Item Table Component

### PR DESCRIPTION
**Pull request by dgcohen**
[Ticket Link](https://jira.nypl.org/browse/SCC-3776)

This PR adds the ItemTableData model and the following components:

ItemTable
RequestButtons
ItemAvailability (formerly InformationLinks)

It also includes tests and some small miscellaneous changes to Item model, SearchResultsBib model, and one renaming change (pascal-casing the itemFilterData class).

NB:
There are some further optimizations to be made, but I will not let these hold up the PR. I added comments where relevant.

### CODE REVIEW

File: src/components/SearchResult/SearchResult.tsx [hunk 1]
```diff
@@ -29,12 +46,19 @@ const SearchResult = ({ bib }: SearchResultProps) => {
         <RCLink href={`${PATHS.BIB}/${bib.id}`}>{bib.title}</RCLink>
       </CardHeading>
       <CardContent>
-        <Box sx={{ p: { display: "inline", marginRight: "s" } }}>
+        <Box sx={{ p: { display: "inline", marginRight: "s" } }} mb="m">
           {bib.materialType && <Text>{bib.materialType}</Text>}
```
Comment (EdwinGuzman) [inline src/components/SearchResult/SearchResult.tsx:50]: Taking a better/closer look, I think these should be spans.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
The update makes sense to me. There are some minor comments but I know this will go through another iteration.

Some things that stick out that we have talked about:
- Use the the "text" variant for the button that looks like a link for now. I think we may need to reword it and isolate it to better appear as a button. This is for the "contact a librarian" button.
- Right now, `Button`s styled as link do have the underlined as hover and it is expected.

Something I noticed after we chatted:
- the mobile view has many horizontal lines that I will talk to @bigfishdesign13 about. It makes the mobile view confusing (to me at least)
Screen Shot 2023-12-04 at 12 38 36 PM

A few things that need to be added that can be another ticket/PR:
- the "View all X items > " link when there are more than 3 items in a search result
- the eResource links that should appear as "Available Online" (https://www.nypl.org/research/research-catalog/search?q=national%20geographic)

Screen Shot 2023-12-04 at 12 40 10 PM

File: src/components/RCLink/RCLink.tsx [hunk 2]
```diff
@@ -1,4 +1,7 @@
-import { Link as DSLink } from "@nypl/design-system-react-components"
+import {
+  Link as DSLink,
+  type LinkTypes,
```
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply src/components/RCLink/RCLink.tsx:3]: It was pointed out that `LinkProps` needs to be exported.

**Comment by bigfishdesign13** [comment]
> * the mobile view has many horizontal lines that I will talk to @bigfishdesign13 about. It makes the mobile view confusing (to me at least)

@EdwinGuzman  Those lines are there to provide delineation between table rows.  But yes, that treatment isn't working well here.  Perhaps we could tie that style to the `showRowDividers` prop.  MLN is the only NYPL app that is using the `Table` component and the horizonal lines in the mobile view don't look great on that site either.

File: ENVIRONMENTVARS.md [hunk 3]
```diff
@@ -20,9 +20,12 @@ If an environment variable is updated, make sure to restart the server for the a
...
+| `NEXT_PUBLIC_CLOSED_LOCATIONS`           | string | "all;Library of the Performing Arts" | A semicolon-delimited list of strings. Include quotes around the string. All locations beginning with any string in this list will be removed from the list of request options in the `ElectronicDelivery`, `HoldRequest`, and `ItemTableRow` components. Currently used physical locations: `Schwarzman;Science;Library for the Performing Arts;Schomburg`. To close all locations, add `all`. This will also remove EDD as a request option, the 'Request' buttons, and also disable the hold request/edd forms. If `all` is not present, EDD and 'Request' buttons will still be available. |
```
Comment (charmingduchess) [inline ENVIRONMENTVARS.md:26]: What does `NEXT_PUBLIC` indicate?

File: src/components/ItemTable/ItemAvailability.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  item: Item
+}
+
+/**
+ * The ItemAvailability component appears below the Item table and displays
+ * info about an item's availability.
+ * TODO: Add Feedback box, Due date, Available font styles
+ */
+const ItemAvailability = ({ item }: ItemAvailabilityProps) => {
+  // TODO: Move this logic into a getter function in the Item class that returns an availability status key
```
Comment (charmingduchess) [inline src/components/ItemTable/ItemAvailability.tsx:16]: Is this follow on work ticketed?

Comment (charmingduchess) [reply src/components/ItemTable/ItemAvailability.tsx:16]: This was always a really dicey area. A status key and switch sounds like a great solution.

File: ENVIRONMENTVARS.md (hunk 3 above)
Comment (dgcohen) [reply ENVIRONMENTVARS.md:26]: That prefix allows env variables to be used on the client side.

File: src/components/ItemTable/ItemAvailability.tsx (hunk 4 above)
Comment (dgcohen) [reply src/components/ItemTable/ItemAvailability.tsx:16]: Just ticketed the follow-up work. Not a blocker for this PR.

https://jira.nypl.org/browse/SCC-3906

File: src/models/SearchResultsBib.ts [hunk 5]
```diff
@@ -33,6 +35,7 @@ export default class SearchResultsBib {
       ? result.publicationStatement[0]
       : null
     this.electronicResources = result.electronicResources || null
+    this.issuance = (result.issuance?.length && result.issuance[0]) || null
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply src/models/SearchResultsBib.ts]: Good catch, just updated the constructor to populate the issuance field with the full array if present and added the "some" check to isArchiveCollection()

File: src/components/ItemTable/ItemAvailability.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    return (
+      <Text>
+        <Box as="span" color="ui.warning.primary">
+          Not available
+        </Box>
+        {item.dueDate && ` - In use until ${item.dueDate}`}
+        {" - Please "}
+        <Button
+          id="contact-librarian"
+          buttonType="link"
```
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply src/components/ItemTable/ItemAvailability.tsx:79]: Does this still need to be addressed?

**Review by EdwinGuzman — COMMENTED** [review]
The update looks good but there are a few things to cover before moving on.

- There are general spacing issues that I think will be picked up during VQA. I think now is a good time to share the Vercel Preview link in the Jira ticket with Apoorva and Clare for their VQA and accessibilty review.
- The electronic resource link seems better because the link text is actual text and not the URL. Do verify that this is what we want (also note there could be some space above the "Available Online" text):
Screen Shot 2023-12-13 at 11 32 56 AM

- The "Not available - Please contact a librarian for assistance" text is not displaying. There also seems to be more spacing between elements and it could be slightly reduced. It also appears that the buttons that should be disabled are not. The right side is from this PR and the search term is "National Geographic":
Screen Shot 2023-12-13 at 11 32 15 AM

**Review by EdwinGuzman — APPROVED** [review]
Approving knowing there's stylistic updates todo later.

**Pull request merged**